### DISCUSSION TITLE
Ui framework is too complex to understand, is it necessary to make the ui that flexible？

### DISCUSSION THREAD

**Discussion by Snailya** [iTwin/itwinjs-core#3478] [Q&A]
For example, when customizing ui, one could define many frontstages in an app, then add these frontstage to fronstage manager, finally call setActiveFrontstageDef to make that frontstage visible.
As for creating fronstage, there are two approach: subclassing StandardFrontstageProvider or FrontStageProvider or create a custom class with a register method to call UiItems register method inside.
As for add frontstage provider, one can call method in ConfigurableUiManager or FronstageManager.
Besides, I'm not sure is there any scenario that many frontstages are needed then to be switched based on user interaction in one application.

**Comment by bsteinbk** [comment]
The latest approach, that only works with the lastest version of the UI (UI 2.0/AppUi), is the StandardFrontstageProvider approach which basically sets up a frontstage where only the content is defined in the FrontstageProvider and all the UI items that populate toolbars, statusbar, and widget panels for that and other stages are defined via one or more UiItemProviders. This allows task-specific UiItemProviders to be created and shared between different stages and different applications that use AppUi.

**Comment by NancyMcCallB** [comment]
Hi, @Snailya ,

Honestly, we agree that the process of creating an AppUi app is complicated. In fact, some of the things you mention, like the StandardFrontstageProvider, were added to simplify the process. The older ones remain for backward compatibility.

Here are some thoughts:

1. You are not required to use multiple frontstages if your app does not need them. A frontstage can be thought of as a task-based page that allows you to filter the available UI to a specific task. It's useful if you have a lot of functionality in your app and you can break it down in this way. For a simple app, you can fit all the necessary UI comfortably in a single stage. If you can't think of a scenario in which you need to filter an app's functionality, don't bother with it.

2. We created the [iTwin Viewer](https://www.npmjs.com/package/@itwin/web-viewer-react) package to simplify app startup. It creates a basic frontstage for you and you only supply the UI you need. It's really the best option for an app that is not so complex that it needs a lot of task organization.

Thanks for your feedback and let us know if you have more questions!
